File: week1_baseline/ruby/11_tui/lib/boukensha/mcp_servers/shell_server.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# MCP server exposing the run_command shell tool. Run as a stdio subprocess
# by lib/boukensha/tools/shell.rb — configuration comes from environment
# variables (root dir, timeout, allowed-commands list) rather than a CLI
# argument, since that's how the parent process (mcp_client.rb) launches it.
#
# Tool logic here is identical to the pre-MCP tools/shell.rb implementation
# — only the transport changed.

require "bundler/setup" if ENV["BUNDLE_GEMFILE"]
require "mcp"
require "open3"
require "timeout"

ROOT    = File.expand_path(ENV.fetch("BOUKENSHA_MCP_SHELL_ROOT"))
TIMEOUT = Integer(ENV.fetch("BOUKENSHA_MCP_SHELL_TIMEOUT", "30"))

# Same three-state env-var convention as Python's shell_server.py: unset =
# allow all; set-but-empty = empty list (reject everything); "a,b,c" = that
# allow-list.
raw_allowed      = ENV["BOUKENSHA_MCP_SHELL_ALLOWED"]
ALLOWED_COMMANDS = raw_allowed.nil? ? nil : raw_allowed.split(",").reject(&:empty?)

allowed_note = ALLOWED_COMMANDS ? " Allowed executables: #{ALLOWED_COMMANDS.join(', ')}." : ""

def text_response(str)
  MCP::Tool::Response.new([{ type: "text", text: str }])
end

server = MCP::Server.new(name: "boukensha-shell")

server.define_tool(
  name: "run_command",
  description: "Run a shell command inside the working directory and return its combined stdout+stderr output. " \
               "Commands run with a #{TIMEOUT}-second timeout.#{allowed_note}",
  input_schema: { properties: { command: { type: "string", description: "The shell command to execute (e.g. 'ruby script.rb', 'ls -la', 'git status')" } }, required: ["command"] }
) do |server_context:, **args|
  command = args[:command]

  if ALLOWED_COMMANDS
    executable = command.to_s.strip.split(/\s+/).first.to_s
    unless ALLOWED_COMMANDS.include?(executable)
      next text_response("error: '#{executable}' is not in the allowed-commands list (#{ALLOWED_COMMANDS.join(', ')})")
    end
  end

  stdout_err, status = nil, nil
  begin
    Timeout.timeout(TIMEOUT) { stdout_err, status = Open3.capture2e(command, chdir: ROOT) }
  rescue Errno::ENOENT => e
    next text_response("error: command not found: #{e.message}")
  rescue Timeout::Error
    next text_response("error: command timed out after #{TIMEOUT}s: #{command}")
  end

  exit_note = status.success? ? "" : "\n[exit #{status.exitstatus}]"
  output    = stdout_err.to_s.strip
  text_response(output.empty? ? "(no output)#{exit_note}" : "#{output}#{exit_note}")
end

MCP::Server::Transports::StdioTransport.new(server).open

File: week3_capable/python/boukensha/mcp_servers/shell_server.py
"""MCP server exposing the run_command shell tool. Run as a stdio
subprocess by boukensha/tools/shell.py — configuration comes from
environment variables (root dir, timeout, allowed-commands list) rather
than a CLI argument or config file, since that's how the parent process
(boukensha/mcp_client.py's McpClient) launches it.

Tool logic here is identical to the pre-MCP boukensha/tools/shell.py
implementation — only the transport changed.
"""

import os
import subprocess
from typing import Annotated

from pydantic import Field

from mcp.server.fastmcp import FastMCP

ROOT = os.path.abspath(os.environ["BOUKENSHA_MCP_SHELL_ROOT"])
TIMEOUT = int(os.environ.get("BOUKENSHA_MCP_SHELL_TIMEOUT", "30"))

# None (env var unset) = allow all; "" (present but empty) = empty list,
# which rejects every command; "cmd1,cmd2" = that allow-list.
_raw_allowed = os.environ.get("BOUKENSHA_MCP_SHELL_ALLOWED")
ALLOWED_COMMANDS = None if _raw_allowed is None else [c for c in _raw_allowed.split(",") if c.strip()]

mcp = FastMCP("boukensha-shell")


def _oops(msg):
    return f"error: {msg}"


_allowed_note = f" Allowed executables: {', '.join(ALLOWED_COMMANDS)}." if ALLOWED_COMMANDS else ""


@mcp.tool(
    description=(
        "Run a shell command inside the working directory and return its combined stdout+stderr "
        f"output. Commands run with a {TIMEOUT}-second timeout.{_allowed_note}"
    )
)
def run_command(
    command: Annotated[str, Field(description="The shell command to execute (e.g. 'python3 script.py', 'ls -la', 'git status')")],
) -> str:
    if ALLOWED_COMMANDS is not None:
        tokens = str(command).strip().split()
        executable = tokens[0] if tokens else ""
        if executable not in ALLOWED_COMMANDS:
            return _oops(f"'{executable}' is not in the allowed-commands list ({', '.join(ALLOWED_COMMANDS)})")

    try:
        result = subprocess.run(
            command, shell=True, cwd=ROOT,
            stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
            timeout=TIMEOUT,
        )
    except subprocess.TimeoutExpired:
        return _oops(f"command timed out after {TIMEOUT}s: {command}")
    except OSError as e:
        return _oops(f"command not found: {e}")

    output = result.stdout.decode("utf-8", errors="replace").strip()
    exit_note = "" if result.returncode == 0 else f"\n[exit {result.returncode}]"
    return f"(no output){exit_note}" if not output else f"{output}{exit_note}"


if __name__ == "__main__":
    mcp.run(transport="stdio")
